### PR TITLE
feat(hatchery): add kubernetes hatchery

### PR DESCRIPTION
**Pull request by bnjjj**
Need more test with a real kubernetes cluster (not in local)

### CODE REVIEW

File: engine/hatchery/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+	"os"
+	"strings"
+	"time"
+
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/cdsclient"
+	"github.com/ovh/cds/sdk/hatchery"
+	"github.com/ovh/cds/sdk/log"
+
+	"github.com/moby/moby/pkg/namesgenerator"
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: don't we have our own namesgenerator ?

File: engine/hatchery/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	if hconfig.API.HTTP.URL == "" {
+		return fmt.Errorf("API HTTP(s) URL is mandatory")
+	}
+
+	if hconfig.API.Token == "" {
+		return fmt.Errorf("API Token URL is mandatory")
+	}
+
+	if hconfig.Name == "" {
+		return fmt.Errorf("please enter a name in your local hatchery configuration")
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: k8s

File: engine/hatchery/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+	return h.client
+}
+
+//Configuration returns Hatchery CommonConfiguration
+func (h *HatcheryKubernetes) Configuration() hatchery.CommonConfiguration {
+	return h.Config.CommonConfiguration
+}
+
+// ModelType returns type of hatchery
+func (*HatcheryKubernetes) ModelType() string {
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: this should be the type of worker model ?

File: engine/hatchery/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+	return sdk.Kubernetes
+}
+
+// CanSpawn return wether or not hatchery can spawn model.
+// requirements are not supported
+func (h *HatcheryKubernetes) CanSpawn(model *sdk.Model, jobID int64, requirements []sdk.Requirement) bool {
+	for _, r := range requirements {
+		if r.Type == sdk.ServiceRequirement {
+			return false
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: i'm so sad

File: engine/hatchery/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+			RestartPolicy:                 apiv1.RestartPolicyNever,
+			TerminationGracePeriodSeconds: &gracePeriodSecs,
+			Containers: []apiv1.Container{
+				{
+					Name:  name,
+					Image: spawnArgs.Model.Image,
+					Env:   envs,
+					Resources: apiv1.ResourceRequirements{
+						Requests: apiv1.ResourceList{
+							apiv1.ResourceMemory: resource.MustParse(fmt.Sprintf("%d", h.Config.DefaultMemory)),
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: we could handle the memory requirement ?

File: engine/hatchery/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+// SpawnWorker starts a new worker process
+func (h *HatcheryKubernetes) SpawnWorker(spawnArgs hatchery.SpawnArguments) (string, error) {
+	name := fmt.Sprintf("k8s-%s-%s", strings.ToLower(spawnArgs.Model.Name), strings.Replace(namesgenerator.GetRandomName(0), "_", "-", -1))
+	label := "execution"
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: why "execution" ?

File: engine/hatchery/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+		},
+	})
+
+	return pod.Name, err
+}
+
+// WorkersStarted returns the number of instances started but
+// not necessarily register on CDS yet
+func (h *HatcheryKubernetes) WorkersStarted() int {
+	list, err := h.k8sClient.CoreV1().Pods(h.Config.KubernetesNamespace).List(metav1.ListOptions{LabelSelector: LABEL_WORKER})
```
Comment (fsamin) [inline engine/hatchery/kubernetes/kubernetes.go]: should we list by the hatchery name (and set the hatchery name as a label) ?

File: engine/hatchery/kubernetes/kubernetes.go (hunk 1 above)
Comment (yesnault) [reply engine/hatchery/kubernetes/kubernetes.go]: hum, not since migrate to dep :-/

File: engine/hatchery/kubernetes/kubernetes.go (hunk 4 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: First iteration

File: engine/hatchery/kubernetes/kubernetes.go (hunk 6 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: Not really inspired :p any suggestion ?

File: engine/hatchery/kubernetes/kubernetes.go (hunk 5 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: Yes i forgot

**Comment by bnjjj** [comment]
No @fsamin deps aren’t already there. I don’t want to put fire to my train’s wifi 😬

File: engine/hatchery/kubernetes/kubernetes.go (hunk 7 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: Yes I can add another label for hatchery name

File: engine/hatchery/kubernetes/kubernetes.go (hunk 1 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: :(

Comment (fsamin) [reply engine/hatchery/kubernetes/kubernetes.go]: don't we have our own namesgenerator package ?

**Comment by bnjjj** [comment]
F**** DEP !!!

File: engine/hatchery/kubernetes/kubernetes.go (hunk 3 above)
Comment (bnjjj) [reply engine/hatchery/kubernetes/kubernetes.go]: done

File: engine/hatchery/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+		}
+	}
+
+	return workersLen
+}
+
+// WorkersStartedByModel returns the number of instances of given model started but
+// not necessarily register on CDS yet
+func (h *HatcheryKubernetes) WorkersStartedByModel(model *sdk.Model) int {
+	list, err := h.k8sClient.CoreV1().Pods(h.Config.KubernetesNamespace).List(metav1.ListOptions{LabelSelector: LABEL_WORKER_MODEL})
```
Comment (yesnault) [inline engine/hatchery/kubernetes/kubernetes.go]: this func returns "WorkersStartedByModel returns the number of instances of given model", but model arg is never used.

**Pull request merged**